packages/amplify_api/lib/src/graphql/graphql_request_factory.dart

```diff
@@ -0,0 +1,135 @@
+    ModelSchema schema = provider.modelSchemas.firstWhere(
+        (elem) => provider.getModelTypeByModelName(elem.name) == modelType,
+        orElse: () => throw ApiException(
+            'No schema found for the ModelType provided',
+            recoverySuggestion:
+                'Pass in a valid modelProvider instance while instantiating APIPlugin or provide a valid ModelType'));
+
+    if (schema.fields == null) {
+      throw ApiException('Schema found does not have a fields property',
+          recoverySuggestion:
+              'Pass in a valid modelProvider instance while instantiating APIPlugin');
+    }
+
+    if (operation == GraphQLRequestOperation.list &&
+        schema.pluralName == null) {
+      throw ApiException('No schema name found',
+          recoverySuggestion:
+              'Pass in a valid modelProvider instance while instantiating APIPlugin or provide a valid ModelType');
+    }
+
+    return schema;
+  }
+
+  /*
+   *  buildQuery()
+   *
+   *  
+   *  Example: 
+   *    query getBlog($id: ID!) { getBlog(id: $id) { id name createdAt } }
+  */
+  GraphQLRequest<T> buildQuery<T extends Model>(
+      {required ModelType modelType,
+      required Map<String, String>? variableInput,
+      required String? id,
+      required GraphQLRequestType requestType,
+      required GraphQLRequestOperation requestOperation}) {
+    ModelSchema schema = _getAndValidateSchema(modelType, requestOperation);
+
+    String name = schema.name;
+
+    // fields to retrieve, e.g. "id name createdAt"
+    String fields = _getFieldsFromModelType(schema);
+
+    // e.g. "query"
+    String requestTypeVal = describeEnum(requestType.toString());
+    // e.g. "get"
+    String requestOperationVal = describeEnum(requestOperation.toString());
+
+    // Upper document input: ($id: ID!)
+    var varInputUpperStr = '';
+    // Lower document input: (id: $id)
+    var varInputLowerStr = '';
+
+    if (variableInput != null && variableInput.isNotEmpty) {
+      List<String> variableInputUpper = [];
+      List<String> variableInputLower = [];
+
+      variableInput.forEach((key, value) {
+        variableInputUpper.add("\$$key: $value!");
+        variableInputLower.add("$key: \$$key");
+      });
+
+      varInputUpperStr = "(${variableInputUpper.join(", ")})";
+      if (requestOperation == GraphQLRequestOperation.get) {
+        varInputLowerStr = "(${variableInputLower.join(", ")})";
+      } else {
+        varInputLowerStr = "(input: { ${variableInputLower.join(", ")} })";
+      }
+    }
+
+    if (requestOperation == GraphQLRequestOperation.list) {
+      name = schema.pluralName!;
+      fields = 'items { $fields } nextToken';
+    }
+
+    String doc =
+        '''$requestTypeVal $requestOperationVal$name${varInputUpperStr} { $requestOperationVal$name$varInputLowerStr { $fields } }''';
+
+    // TODO: create input map for variables & connect with current variableInput
+    Map<String, dynamic>? variables = id != null ? {"id": id} : {};
```

Comment:
nit: I don't think this variable is nullable bc you are assigning it 1 of 2 things, so can be `Map<String, String>` I think, but also looks like you are gonna work on this soon, anyway.